src/cfg/structured.rs

```diff
@@ -91,3 +105,131 @@ impl StructuredBlock {
         }
     }
 }
+
+impl StructuredFunction {
+    pub fn to_function(&self) -> Function {
+        let mut builder = StructuredCfgBuilder {
+            code: vec![],
+            scopes: vec![],
+            name_counter: 0,
+        };
+        self.block.to_code(&mut builder);
+
+        Function {
+            name: self.name.clone(),
+            args: self.args.clone(),
+            instrs: builder.code,
+            pos: None,
+            return_type: None,
+        }
+    }
+}
+
+pub struct StructuredCfgBuilder {
+    code: Vec<Code>,
+    scopes: Vec<String>,
+    name_counter: usize,
+}
+
+impl StructuredCfgBuilder {
+    fn fresh(&mut self) -> String {
+        self.name_counter += 1;
+        format!("sblock___{}", self.name_counter - 1)
+    }
+
+    fn scope_break_to(&self, num: i64) -> String {
+        assert!(num > 0);
+        self.scopes[self.scopes.len() - num as usize].clone()
+    }
+}
+
+impl StructuredBlock {
+    pub(crate) fn to_code(&self, builder: &mut StructuredCfgBuilder) {
+        match self {
+            StructuredBlock::Basic(block) => builder.code.extend(block.to_code()),
+            StructuredBlock::Block(block) => {
+                let block_end_name = builder.fresh();
+                builder.scopes.push(block_end_name.clone());
+                block.to_code(builder);
+                builder.code.push(Code::Label {
+                    label: block_end_name.clone(),
+                    pos: None,
+                });
+                assert!(builder.scopes.pop().unwrap() == block_end_name);
```

Comment:
intentional to remove the scope you just added? why?

Reply:
added comment- we pop the scope after recurring and thus finishing the block